### PR TITLE
Maven global settings and toolchains UI cleanup

### PR DESCRIPTION
**Pull request by G-Ork**
I redesigned the Maven user settings and add the sections _Global_ & _User._  For booth there are now _settings_ & _toolchains_ options. 

- Added default path hints for global settings and global toolchains. They are only rendered for external maven installations. Don't forget to apply changes to see them.
- Unify settings override through goals field. All four settings could be put into the goal in short form (-gt. -gs, -t, -s).
- Throw exception during launch if path to a settings file is invalid.
- Provide unit-tests for maven launcher for all four settings
- Fix a bug in settings validation. Parser problems are not copied to result if parser do not crash. All well formed xml-files are treated as valid.
- Add XSDs for settings and toolchains as eclipse catalog contributions. 
  -  Could not tests this as i am not able to start M2E-IDE this the _WST XML_ feature.

### CODE REVIEW

**Comment by laeubi** [comment]
> Add XSDs for settings and toolchains as eclipse catalog contributions.
> 
>     Could not tests this as i am not able to start M2E-IDE this the WST XML feature.

@mickaelistria @fbricon can you probably help here?

**Comment by laeubi** [comment]
@G-Ork can you share a screenshot before/after?

**Comment by G-Ork** [comment]
@laeubi Holy moly fast. Explain more what kind of screenshot you like.

**Comment by laeubi** [comment]
> I redesigned the Maven user settings and add the sections Global & User. For booth there are now settings & toolchains options.

So what ever you redesigned in the UI, e.g dialog before your change, and how it looks now :-)

File: org.eclipse.m2e.core.tests/src/org/eclipse/m2e/core/IBiConsumer.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ *
+ * @param <T> the type of the first argument to the operation
+ * @param <U> the type of the second argument to the operation
+ * 
+ * @see Consumer
+ * @author Georg Tsakumagos
+ */
+@FunctionalInterface
+public interface IBiConsumer<T, U> {
```
Comment (laeubi) [inline org.eclipse.m2e.core.tests/src/org/eclipse/m2e/core/IBiConsumer.java:34]:
```suggestion
public interface CoreBiConsumer<T, U> {
```

**Comment by laeubi** [comment]
@G-Ork you also need to consider two other places, one is

`org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.createExecutionRequest(IMavenConfiguration, IComponentLookup, MavenSettingsLocations, File)`

where we need to set the Global/User Toolchain file initally, then in `org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.newExecutionRequest()` you probabbly need to update them if the are configured in the `MavenProperties` differently

**Comment by G-Ork** [comment]
Okay i was not exact. I am able to start the M2E-IDE but the WST XML feature are not loaded. Here is the screenshot based on this PR.

**Current (released)**
Maven-Usersettings-Released

**Draft from PR**
Maven-Usersettings-Draft

File: org.eclipse.m2e.core.tests/src/org/eclipse/m2e/core/IBiConsumer.java (hunk 1 above)
Comment (G-Ork) [reply org.eclipse.m2e.core.tests/src/org/eclipse/m2e/core/IBiConsumer.java:34]: Okay sounds better. Same applies to the supplier.

**Comment by G-Ork** [comment]
> @G-Ork you also need to consider two other places, one is
> 
> where we need to set the Global/User Toolchain file initally, then in `org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.newExecutionRequest()` you probabbly need to update them if the are configured in the `MavenProperties` differently

@laeubi From what i can see it your feature is not complete. The [.mvn/maven.config ](https://maven.apache.org/configure.html#mvn-maven-config-file) can contain much more than just [alternative settings](https://github.com/eclipse-m2e/m2e-core/blob/93f9495666d3d2f32433093d9b46aec845c1eaee/org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/embedder/MavenExecutionContext.java#L164) location. 

Any chance we can refactor this to be more easy and support more options like .e.g. "-T3 -U "?
As far i can see it should work because you only care for the settings file. This part uses the configuration pointing to the settings files.

**Comment by laeubi** [comment]
> From what i can see it your feature is not complete.

That's why I said you probably want to enhance it to support toolchains

> Any chance we can refactor this to be more easy and support more options like .e.g. "-T3 -U "?

There is no easy way as not all options can be supported and probably need to be supported differently. For example `-T` does not makes any sense as m2e only ever uses one build thread internally so claiming to use more might even be confusing. `-U` looks easy enough at a first glance, but this is usually controlled by the UI so when should we set it?

So I think one should really focus on things one by one when there is a demand.

**Comment by HannesWell** [comment]
@G-Ork thanks for this contribution.
Since you listed many points in your intial comment I wonder if it would be possible (without too much effort) to split this into multiple PRs? This would also make it easier to review.
Furthermore for contributions of non-Committers over 1000LOC we would actually need an separate IP-clearance (usually not denied, but extra work).

I guess that the UI refactoring and the core changes can probably be splitted, similar to the schema additions.

**Comment by G-Ork** [comment]
@HannesWell The high LOC includes the xsd files which are IP from maven. I think we could split XSD changes from this one into a separate PR. The changes for UI are so small i would not exclude them from the unit-tests and m2e layers.

**Comment by HannesWell** [comment]
> @HannesWell The high LOC includes the xsd files which are IP from maven. I think we could split XSD changes from this one into a separate PR. The changes for UI are so small i would not exclude them from the unit-tests and m2e layers.

Sounds good. I have to admit that I first have to recall for what the XSD files are actually used.

**Comment by G-Ork** [comment]
Split PR into two separate and close this:

- #1691 
- #1690

**Comment by G-Ork** [comment]
Split PR into two separate and close this:

- #1691 
- #1690

**Pull request closed without merging**